File: app/src/main/java/com/stho/software/nyota/views/AbstractSkyView.java
package com.stho.software.nyota.views;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PointF;
import android.util.AttributeSet;
import android.view.DragEvent;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.view.View;

import com.stho.software.nyota.universe.Constellation;
import com.stho.software.nyota.utilities.Degree;
import com.stho.software.nyota.universe.IElement;
import com.stho.software.nyota.ISkyActivity;
import com.stho.software.nyota.utilities.Point;
import com.stho.software.nyota.data.Settings;
import com.stho.software.nyota.utilities.SphereProjection;
import com.stho.software.nyota.universe.Star;
import com.stho.software.nyota.utilities.Topocentric;

import org.jetbrains.annotations.Nullable;

import java.util.HashMap;

abstract class AbstractSkyView extends View implements View.OnDragListener {
    private final Paint green = new Paint();
    private final Paint white = new Paint();
    private final Paint yellow = new Paint();
    private final Paint gray = new Paint();
    private final Paint blue = new Paint();
    private final Paint orange = new Paint();
    private final Path path = new Path();
    private HashMap<Star, PointF> positions = new HashMap<>();
    private HashMap<Integer, Bitmap> bitmaps = new HashMap<>();
    protected boolean displayNames = true;
    protected boolean displaySymbols = true;
    protected boolean displayMagnitude = true;
    protected Topocentric center = new Topocentric(0,0);
    protected ScaleGestureDetector scaleGestureDetector;
    protected GestureDetector gestureDetector;
    private double zoomAngle = 45;
    private SphereProjection projection = new SphereProjection();
    private boolean isScrollingEnabled = true;
    private boolean isScalingEnabled = true;
    private boolean drawGrid = true;
    private ISkyActivity activity = null;

    public AbstractSkyView(Context context, AttributeSet attrs) {
        super(context, attrs);
        onCreate(context);
    }

    public void register(ISkyActivity activity) {
        this.activity = activity;
    }

    void setScrollingEnabled(boolean enable) {
        isScrollingEnabled = enable;
    }

    void setScalingEnabled(boolean enable) {
        isScalingEnabled = enable;
    }

    private void onCreate(final Context context) {
        loadSettings();
        scaleGestureDetector = new ScaleGestureDetector(context, new ScaleGestureDetector.SimpleOnScaleGestureListener() {
            @Override
            public boolean onScale(ScaleGestureDetector detector) {
                applyScale(detector.getScaleFactor());
                return true;
            }
        });
        gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
            @Override
            public boolean onDoubleTap(MotionEvent e) {
                resetTransformation();
                return false;
            }
            @Override
            public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
                applyScrolling(distanceX, distanceY);
                return false; // super.onScroll(e1, e2, distanceX, distanceY);
            }

            @Override
            public boolean onSingleTapConfirmed(MotionEvent e) {
                if (activity != null) {
                    return activity.onSingleTapConfirmed(e);
                }
                return super.onSingleTapConfirmed(e);
            }
        });
        yellow.setColor(Color.YELLOW);
        yellow.setAlpha(120);
        yellow.setStyle(Paint.Style.STROKE);
        green.setColor(Color.GREEN);
        green.setAlpha(200);
        green.setStyle(Paint.Style.FILL_AND_STROKE);
        green.setAntiAlias(true);
        green.setTextSize(50);
        white.setColor(Color.WHITE);
        white.setAlpha(255);
        white.setStyle(Paint.Style.FILL_AND_STROKE);
        white.setAntiAlias(true);
        gray.setColor(Color.GRAY);
        gray.setAlpha(200);
        gray.setStyle(Paint.Style.FILL_AND_STROKE);
        gray.setAntiAlias(true);
        gray.setTextSize(40);
        blue.setColor(Color.BLUE);
        blue.setAlpha(200);
        blue.setStyle(Paint.Style.FILL_AND_STROKE);
        blue.setAntiAlias(true);
        blue.setTextSize(40);
        orange.setColor(Color.rgb(253, 106, 2));
        orange.setAlpha(120);
        orange.setStyle(Paint.Style.FILL_AND_STROKE);
        orange.setAntiAlias(true);
        orange.setTextSize(40);
    }

    private void loadSettings() {
        Settings settings = new Settings();
        settings.load(this.getContext());
        onLoadSettings(settings);
    }

    protected void onLoadSettings(Settings settings) {
        // nothing. overwrite ...
    }

    void raiseOnChangeSkyCenter() {
        if (activity != null) {
            activity.onChangeSkyCenter();
        }
    }

    void applyScrolling(double dx, double dy) {
        if (isScrollingEnabled) {
            double zoom = getZoom();
            center.azimuth += Degree.arcTan2(dx, zoom);
            center.altitude -= Degree.arcTan2(dy, zoom);
            raiseOnChangeSkyCenter();
            updateUI();
        }
    }

    public void applyScale(double scaleFactor) {
        if (isScalingEnabled) {
            final double MIN_ZOOM_ANGLE = 0.5;
            final double MAX_ZOOM_ANGLE = 120;
            zoomAngle = Math.max(MIN_ZOOM_ANGLE, Math.min(MAX_ZOOM_ANGLE, zoomAngle / scaleFactor));
            updateUI();
        }
    }

    void resetTransformation() {
        zoomAngle = 45;
        center.azimuth = getReference().azimuth;
        center.altitude = getReference().altitude;
        raiseOnChangeSkyCenter();
        updateUI();
    }

    public void setZoomAngle(double angle) {
        zoomAngle = angle;
        updateUI();
    }

    @Override
    public boolean onDrag(View view, DragEvent dragEvent) {
        return false;
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        scaleGestureDetector.onTouchEvent(event);
        gestureDetector.onTouchEvent(event);
        return true;
    }

    public void updateUI() {
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        int w = getWidth();
        int h = getHeight();
        canvas.translate(w / 2, h / 2);
        double zoom = getZoom();
        projection.setCenter(center.azimuth, center.altitude);
        positions.clear();
        if (drawGrid) {
            drawGrid(canvas, w, h, zoom);
        }
        onDrawElements(canvas, zoom);
    }

    public Topocentric getCenter() {
        return center;
    }

    abstract protected Topocentric getReference();

    abstract protected void onDrawElements(Canvas canvas, double zoom);

    private double getZoom() {
        int w = getWidth();
        return 0.5 * w / Degree.tan(0.5 * zoomAngle);
    }

    protected void drawStar(Canvas canvas, double zoom, Star star) {
        calculatePosition(zoom, star);
        drawStar(canvas, star, white);
    }

    protected void drawConstellation(Canvas canvas, double zoom, Constellation constellation) {
        calculatePositions(zoom, constellation);
        drawConstellation(canvas, constellation);
    }

    protected void drawSun(Canvas canvas, double zoom, IElement sun) {
        PointF p = getPosition(zoom, sun.getPosition());
        if (isOnScreen(p)) {
            Bitmap bm = getScaledBitmap(sun.getImageId(), 32, 32);
            canvas.drawBitmap(bm, p.x - 16, p.y - 16, white);
        }
    }

    protected void drawMoon(Canvas canvas, double zoom, IElement moon) {
        PointF p = getPosition(zoom, moon.getPosition());
        if (isOnScreen(p)) {
            Bitmap bm = getScaledBitmap(moon.getImageId(), 48, 48);
            canvas.drawBitmap(bm, p.x - 24, p.y - 24, white);
        }
    }

    protected void drawPlanet(Canvas canvas, double zoom, IElement element) {
        PointF p = getPosition(zoom, element.getPosition());
        if (isOnScreen(p)) {
            Bitmap bm = getScaledBitmap(element.getImageId(), 16, 16);
            canvas.drawBitmap(bm, p.x - 8, p.y - 8, white);
        }
    }

    protected void drawName(Canvas canvas, double zoom, IElement element) {
        drawName(canvas, zoom, element.getPosition(), element.getName());
    }

    protected void drawName(Canvas canvas, double zoom, Topocentric position, String name) {
        PointF p = getPosition(zoom, position);
        if (isOnScreen(p)) {
            canvas.drawText(name, p.x + 10, p.y - 10, orange);
        }
    };



    private Bitmap getScaledBitmap(int resourceId, int newWidth, int newHeight) {
        if (bitmaps.containsKey(resourceId)) {
            return bitmaps.get(resourceId);
        } else {
            Bitmap bm = createScaledBitmap(resourceId, newWidth, newHeight);
            bitmaps.put(resourceId, bm);
            return bm;
        }
    }

    private Bitmap createScaledBitmap(int resourceId, int newWidth, int newHeight) {
        Bitmap bm = BitmapFactory.decodeResource(getResources(), resourceId);
        return Bitmap.createScaledBitmap(bm, newWidth, newHeight, false);
    }

    private void calculatePosition(double zoom, Star star) {
        positions.put(star, getPosition(zoom, star.getPosition()));
    }

    private void calculatePositions(double zoom, Constellation constellation) {
        for (Star star : constellation.getStars()) {
            positions.put(star, getPosition(zoom, star.getPosition()));
        }
    }

    private @Nullable PointF getPosition(double zoom, Topocentric topocentric) {
        return getPosition(zoom, topocentric.azimuth, topocentric.altitude);
    }

    private @Nullable PointF getPosition(double zoom, double azimuth, double altitude) {
        Point p = projection.getImagePoint(azimuth, altitude);
        if (p != null) {
            float x = (float) (zoom * p.x);
            float y = (float) (zoom * p.y);
            return new PointF(x, -y);
        }
        return null;
    }

    private void drawConstellation(Canvas canvas, Constellation constellation) {
        for (Star star : constellation.getStars()) {
            drawStar(canvas, star, green);
        }
        for (Star[] line : constellation.getLines()) {
            drawLine(canvas, line);
        }
    }

    private void drawStar(Canvas canvas, Star star, Paint color){
        PointF p = positions.get(star);
        if (isOnScreen(p)) {
            if (displayMagnitude) {
                applyMagnitude(color, star.getBrightness());
            }
            canvas.drawCircle(p.x, p.y, 4, color);
            if (displaySymbols) {
                canvas.drawText(star.getSymbol(), p.x, p.y, gray);
            }
        }
    }

    private void applyMagnitude(Paint color, double magnitude) {
        if (magnitude > 5.0)
            color.setAlpha(100);
        else if (magnitude > 3.0)
            color.setAlpha(150);
        else if (magnitude > 1.0)
            color.setAlpha(200);
        else
            color.setAlpha(255);
    }

    private void drawLine(Canvas canvas, Star[] line) {
        boolean first = true;
        path.reset();
        for (Star star : line) {
            PointF p = positions.get(star);
            if (isOnScreen(p)) {
                if (first) {
                    path.moveTo(p.x, p.y);
                    first = false;
                } else {
                    path.lineTo(p.x, p.y);
                }
            } else {
                first = true;
            }
        }
        canvas.drawPath(path, yellow);
    }

    private void drawGrid(Canvas canvas, int w, int h, double zoom) {
        if (center.altitude > 0) {
            for (float deltaAzimuth = 0; deltaAzimuth < 180; deltaAzimuth += 10) {
                double azimuth = center.azimuth + deltaAzimuth;
                for (float altitude = 85; altitude > -85; altitude -= 5) {
                    PointF p = getPosition(zoom, azimuth, altitude);
                    if (p != null) {
                        if (p.x > w || p.y > h) {
                            break;
                        }
                        if (isOnScreen(p)) {
                            canvas.drawCircle(p.x, p.y, 2, blue);
                            canvas.drawCircle(-p.x, p.y, 2, blue);
                        }
                    }
                }
            }
        } else {
            for (float deltaAzimuth = 0; deltaAzimuth < 180; deltaAzimuth += 10) {
                double azimuth = center.azimuth + deltaAzimuth;
                for (float altitude = -85; altitude < 85; altitude += 5) {
                    PointF p = getPosition(zoom, azimuth, altitude);
                    if (p != null) {
                        if (p.x > w || p.y < -h) {
                            break;
                        }
                        if (isOnScreen(p)) {
                            canvas.drawCircle(p.x, p.y, 2, blue);
                            canvas.drawCircle(-p.x, p.y, 2, blue);
                        }
                    }
                }
            }
        }
    }

    private boolean isOnScreen(PointF p) {
        if (p != null) {
            float w = getWidth();
            float h = getHeight();
            return (Math.abs(p.x) < w && Math.abs(p.y) < h);
        }
        return false;
    }
}

File: app/src/main/java/com/stho/nyota/views/AbstractSkyView.kt
package com.stho.nyota.views

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Path
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.view.*
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.ScaleGestureDetector.SimpleOnScaleGestureListener
import com.stho.nyota.ISkyViewListener
import com.stho.nyota.settings.Settings
import com.stho.nyota.sky.universe.*
import com.stho.nyota.sky.utilities.Degree
import com.stho.nyota.sky.utilities.Topocentric
import com.stho.nyota.sky.utilities.projections.ISphereProjection
import com.stho.nyota.sky.utilities.projections.SphereProjection
import com.stho.nyota.ui.sky.IViewOptions
import java.util.*


abstract class AbstractSkyView(context: Context?, attrs: AttributeSet?): View(context, attrs), View.OnDragListener {

    lateinit var options: IViewOptions
        private set


    private val bitmaps = HashMap<Int, Bitmap>()
    private var projection: ISphereProjection = SphereProjection()
    private var draw: SkyDraw = SkyDraw()

    fun setOptions(options: IViewOptions) {
        this.options = options
        this.draw.touch()
        invalidate()
    }

    fun touch() {
        this.draw.touch()
        invalidate()
    }

    val path = Path()
    val center = Topocentric(0.0, 0.0)
    var zoomAngle: Double = Settings.DEFAULT_ZOOM_ANGLE
        set(value) {
            val validZoomAngle = value.coerceIn(Settings.MIN_ZOOM_ANGLE, Settings.MAX_ZOOM_ANGLE)
            if (field != validZoomAngle) {
                field = validZoomAngle
                touch()
            }
        }

    var scaleGestureDetector: ScaleGestureDetector? = null
    var gestureDetector: GestureDetector? = null
    var tippedPosition: Topocentric? = null

    private var listener: ISkyViewListener? = null

    init {
        onCreate(context)
    }

    val sensitivityAngle: Double
        get() {
            val metrics: DisplayMetrics = resources.displayMetrics
            val dp: Double = metrics.density * 17.0
            return projection.sensitivityAngle * dp
        }

    fun registerListener(listener: ISkyViewListener?) {
        this.listener = listener
    }

    private fun onCreate(context: Context?) {
        scaleGestureDetector =
            ScaleGestureDetector(context, object : SimpleOnScaleGestureListener() {
                override fun onScale(detector: ScaleGestureDetector): Boolean {
                    applyScale(detector.scaleFactor.toDouble())
                    return true
                }
            })
        gestureDetector = GestureDetector(context, object : SimpleOnGestureListener() {
            override fun onDoubleTap(e: MotionEvent): Boolean {
                listener?.apply {
                    onDoubleTap() // shall call: SkyView.resetTransformation()
                }
                return false
            }

            override fun onScroll(e1: MotionEvent, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
                applyScrolling(distanceX.toDouble(), distanceY.toDouble())
                return false // super.onScroll(e1, e2, distanceX, distanceY);
            }

            override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
                listener?.apply {
                    val p = SkyPointF.forMotionEvent(e, width, height)
                    projection.inverseZoomImagePoint(p)?.let {
                        tippedPosition = it
                        onSingleTap(it);

                    }
                }
                return super.onSingleTapConfirmed(e)
            }
        })
    }

    open fun raiseOnChangeCenter() {
        listener?.apply {
            onChangeCenter()
        }
    }

    open fun raiseOnChangeZoom() {
        listener?.apply {
            onChangeZoom()
        }
    }

    open fun applyScrolling(dx: Double, dy: Double) {
        center.azimuth = Degree.normalize(center.azimuth + projection.calculateAngle(dx))
        center.altitude = (center.altitude - projection.calculateAngle(dy)).coerceIn(-90.0, 90.0)
        tippedPosition = null
        raiseOnChangeCenter()
        invalidate()
    }

    open fun applyScale(scaleFactor: Double) {
        zoomAngle /= scaleFactor
        tippedPosition = null
        raiseOnChangeZoom()
    }

    open fun resetTransformation() {
        tippedPosition = null
        zoomAngle = Settings.DEFAULT_ZOOM_ANGLE
        raiseOnChangeZoom()
        referencePosition?.apply {
            center.azimuth = azimuth
            center.altitude = altitude
        }
        raiseOnChangeCenter()
        invalidate()
    }

    override fun onDrag(p0: View?, p1: DragEvent?): Boolean {
        return false
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        scaleGestureDetector?.onTouchEvent(event)
        gestureDetector?.onTouchEvent(event)
        return true
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        canvas.translate(width / 2f, height / 2f)

        ensureProjection()
        ensureDraw(canvas)

        if (options.displayGrid) {
            draw.drawGrid()
        }

        onDrawElements()
    }

    private fun ensureProjection() {
        if (projection.projection != options.sphereProjection) {
            projection = ISphereProjection.create(options.sphereProjection)
        }
        projection.setZoom(zoomAngle, width)
        projection.setCenter(center.azimuth, center.altitude)
    }

    private fun ensureDraw(canvas: Canvas) {
        draw.configure(canvas, width, height, center)
        draw.options = options
        draw.projection = projection
    }

    protected abstract fun onDrawElements()

    protected abstract val referencePosition: Topocentric?

    fun setCenter(position: Topocentric?) =
        position?.apply {
            setCenter(azimuth, altitude)
        }

    fun setCenter(azimuth: Double, altitude: Double) {
        if (azimuth != center.azimuth || altitude != center.altitude) {
            center.azimuth = azimuth
            center.altitude = altitude
            invalidate()
        }
    }

    protected fun drawLight() =
        draw.drawLight()

    protected fun drawEcliptic(ecliptic: Collection<Topocentric>) =
        draw.drawEcliptic(ecliptic)

    protected fun drawSun(sun: IElement) =
        draw.drawSun(sun, getScaledBitmap(sun.imageId, 72, 72))

    protected fun drawMoon(moon: IElement) =
        draw.drawMoon(moon, getScaledBitmap(moon.imageId, 48, 48))

    protected fun drawPlanet(planet: IElement) =
        draw.drawPlanet(planet, getScaledBitmap(planet.imageId, 32, 32))

    protected fun drawTarget(target: com.stho.nyota.sky.universe.Target) =
        draw.drawTarget(target, getScaledBitmap(target.imageId, 16, 16))

    protected fun drawNameOf(element: IElement) =
        draw.drawNameOf(element)

    protected fun drawGalaxy(galaxy: Galaxy, referenceType: ReferenceType) =
        draw.drawGalaxy(galaxy, referenceType)

    protected fun drawZenit(zenit: Topocentric) =
        draw.drawZenit(zenit)

    protected fun drawNadir(nadir: Topocentric) =
        draw.drawNadir(nadir)

    protected fun drawStar(star: Star, referenceType: ReferenceType) =
        draw.drawStar(star, referenceType)

    protected fun drawConstellation(constellation: Constellation, referenceType: ReferenceType) =
        draw.drawConstellation(constellation, referenceType)

    protected fun drawSatellite(satellite: Satellite) =
        draw.drawSatellite(satellite, getScaledBitmap(satellite.imageId, 72, 72))

    protected fun drawName(position: Topocentric, name: String) =
        draw.drawName(position, name)

    protected fun drawHint(hint: Hint) =
        draw.drawHint(hint)

    protected fun drawElement(position: Topocentric, name: String, luminosity: Luminosity) =
        draw.drawElement(position, name, luminosity)

    fun drawSensitivityArea(position: Topocentric) =
        draw.drawSensitivityArea(position, sensitivityAngle)

    /**
    get image from cache or create it
     */
    private fun getScaledBitmap(resourceId: Int, newWidth: Int, newHeight: Int): Bitmap =
        bitmaps[resourceId] ?: createScaledBitmapIntoCache(resourceId, newWidth, newHeight)

    private fun createScaledBitmapIntoCache(resourceId: Int, newWidth: Int, newHeight: Int): Bitmap =
        createScaledBitmap(resourceId, newWidth, newHeight).also {
            bitmaps[resourceId] = it
        }

    private fun createScaledBitmap(resourceId: Int, newWidth: Int, newHeight: Int): Bitmap =
        BitmapFactory.decodeResource(resources, resourceId).let {
            Bitmap.createScaledBitmap(it, newWidth, newHeight, false)
        }
}



